universe/ValueRefs.cpp

```diff
@@ -2015,26 +1775,64 @@ double ComplexVariable<double>::Eval(const ScriptingContext& context) const
 {
     const std::string& variable_name = m_property_name.back();
 
+    std::function<float (const ShipHull&)> hull_property{nullptr};
+
+    if (variable_name == "HullFuel")
+        hull_property = &ShipHull::Fuel;
+    else if (variable_name == "HullStealth")
+        hull_property = &ShipHull::Stealth;
+    else if (variable_name == "HullStructure")
+        hull_property = &ShipHull::Structure;
+    else if (variable_name == "HullSpeed")
+        hull_property = &ShipHull::Speed;
+
+    if (hull_property) {
+        std::string ship_hull_name;
+        if (m_string_ref1)
+            ship_hull_name = m_string_ref1->Eval(context);
+
+        const ShipHull* ship_hull = GetShipHull(ship_hull_name);
+        if (!ship_hull)
+            return 0.0f;
+
+        return hull_property(*ship_hull);
+    }
+
     // empire properties indexed by integers
-    if (variable_name == "PropagatedSystemSupplyRange" ||
-        variable_name == "SystemSupplyRange" ||
-        variable_name == "PropagatedSystemSupplyDistance")
-    {
-        int empire_id = ALL_EMPIRES;
+    std::function<const std::map<int, float>& (const Empire&)> empire_property{nullptr};
+
+    if (variable_name == "PropagatedSystemSupplyRange")
+        empire_property = [](const Empire& empire){ return GetSupplyManager().PropagatedSupplyRanges(empire.EmpireID()); };
+    if (variable_name == "SystemSupplyRange")
+        empire_property = &Empire::SystemSupplyRanges;
+    if (variable_name == "PropagatedSystemSupplyDistance")
+        empire_property = [](const Empire& empire){ return GetSupplyManager().PropagatedSupplyDistances(empire.EmpireID()); };
+
+    if (empire_property) {
+        using namespace boost::adaptors;
+
+        Empire* empire{nullptr};
+
         if (m_int_ref1) {
-            empire_id = m_int_ref1->Eval(context);
+            int empire_id = m_int_ref1->Eval(context);
             if (empire_id == ALL_EMPIRES)
                 return 0.0;
+            empire = GetEmpire(empire_id);
         }
 
-        // if a key integer is specified, get just that entry (for single empire or sum of all empires)
-        if (m_int_ref2) {
-            int key_int = m_int_ref2->Eval(context);
-            return GetFloatEmpirePropertySingleKey(empire_id, variable_name, key_int);
-        }
+        std::function<bool (const std::map<int, float>::value_type&)> key_filter;
+        key_filter = [](auto k){ return true; };
+
+        if (m_int_ref2)
+            key_filter = [k = m_int_ref2->Eval(context)](auto e){ return k == e.first; };
+
+        if (empire)
+            return boost::accumulate(empire_property(*empire) | filtered(key_filter) | map_values, 0.0f);
```

Comment:
Is this clear and beneficial? Do you anticipate some extension in which something other than all-keys and just-one-key are used? If not this seems overly complicated and hard to quickly understand when using `accumulate`.

Reply:
What do you mean?  The general pattern like[ those lines].(https://github.com/freeorion/freeorion/pull/2928/files/17c050391dc6d63e6997b8b78a868dd0984aeaaa#diff-a3119b793e03ce2507143dbe73a6554dR1829-R1835) or this specific line?

> Is this clear and beneficial?

I would argure yes, this is clear and concise. Applying the same pattern for all properties would make it possible to have a generalized implementation, which only requires to bind the property on construction of the ValueRef while having the evaluation doesn't need those massive if else branches anymore.

> Do you anticipate some extension in which something other than all-keys and just-one-key are used? 

No, but it both variants can be handled with the same statement whilst allowing composability whenever the need arises.

> If not this seems overly complicated and hard to quickly understand when using accumulate.

Why exactly?  `accumulate` is a standard library function and should be known by most people. It sums up the content of a container.  This container here are the extracted values of a map, with an optional applied filter by the map key.

Reply:
The general pattern.

The issue I have with `accumulate` is not what it does internally, but rather that it is somewhat misleading when used here with the two cases of "all" and "just one". Accumulating with a filter to pick just one item in the map is not, to me, intuitive, and suggests there is something more complicated going at first glace. The previous implementation was, to me, clearer, in that it only did the loop and sum when all the entries were being summed. Looping over all the entries to test each instead of using the built-in map lookup seems strange (and is slightly inefficient, but likely not enough so to matter).

Presumably you could just do the accumulate step when summing all the values from the map, and skip it when lookup up just one?

The same pattern can be applied for all properties... I don't have any issue with the function object to do the lookup, which has a clear path to future optimization as you've described.

Reply:
> Accumulating with a filter to pick just one item in the map is not, to me, intuitive, and suggests there is something more complicated going at first glace.

Well, then I just have to disagree here.  I don't see what make the one empire case so special that it requires a different code path when it can be handled as a container with `size() == 1`. 

> Presumably you could just do the accumulate step when summing all the values from the map, and skip it when lookup up just one?

Sure I do it that way if it prevents merging this PR.

However your suggestion needs to be adopted for both the "all empires" and "that empire" cases so it's 4 branches for each property type.  I still believe this can be implemented in a less convoluted way.

Reply:
will get back to this...